tests/grpc/config_test.go

```diff
@@ -0,0 +1,525 @@
+            "fizz": "buzz"
+          },
+          "s": "foo"
+        },
+        "s": "foo"
+      }
+    },
+    "response": {
+      "changes": "DIFF_NONE",
+      "hasDetailedDiff": true
+    },
+    "metadata": {
+      "kind": "resource",
+      "mode": "client",
+      "name": "config"
+    }
+  },
+  {
+    "method": "/pulumirpc.ResourceProvider/Configure",
+    "request": {
+      "variables": {
+        "config:config:a": "[\"one\",\"two\"]",
+        "config:config:b": "true",
+        "config:config:db": "true",
+        "config:config:di": "42",
+        "config:config:ds": "defString",
+        "config:config:i": "42",
+        "config:config:m": "{\"fizz\":\"buzz\"}",
+        "config:config:n": "{\"a\":[\"one\",\"two\"],\"b\":true,\"i\":42,\"m\":{\"fizz\":\"buzz\"},\"s\":\"foo\"}",
+        "config:config:s": "foo"
+      },
+      "args": {
+        "a": [
+          "one",
+          "two"
+        ],
+        "b": true,
+        "db": true,
+        "di": 42,
+        "ds": "defString",
+        "i": 42,
+        "m": {
+          "fizz": "buzz"
+        },
+        "n": {
+          "a": [
+            "one",
+            "two"
+          ],
+          "b": true,
+          "i": 42,
+          "m": {
+            "fizz": "buzz"
+          },
+          "s": "foo"
+        },
+        "s": "foo"
+      },
+      "acceptSecrets": true,
+      "acceptResources": true
+    },
+    "response": {
+      "acceptSecrets": true,
+      "supportsPreview": true,
+      "acceptResources": true,
+      "acceptOutputs": true
+    },
+    "metadata": {
+      "kind": "resource",
+      "mode": "client",
+      "name": "config"
+    }
+  }
+]`
+	replayConfig(t, sequence)
+}
+
+type Config struct {
+	String string            `pulumi:"s"`
+	Bool   bool              `pulumi:"b"`
+	Int    int               `pulumi:"i"`
+	Map    map[string]string `pulumi:"m"`
+	Arr    []string          `pulumi:"a"`
+	Nested ConfigNested      `pulumi:"n"`
+
+	DString string `pulumi:"ds,optional"`
+	DBool   *bool  `pulumi:"db,optional"`
+	DInt    int    `pulumi:"di,optional"`
+}
+
+type ConfigNested struct {
+	String string            `pulumi:"s"`
+	Bool   bool              `pulumi:"b"`
+	Int    int               `pulumi:"i"`
+	Map    map[string]string `pulumi:"m"`
+	Arr    []string          `pulumi:"a"`
+}
+
+var _ = (infer.Annotated)((*Config)(nil))
+
+func (c *Config) Annotate(a infer.Annotator) {
+	a.SetDefault(&c.String, nil, "STRING")
+	a.SetDefault(&c.Bool, nil, "BOOL")
+	a.SetDefault(&c.Int, nil, "INT")
+
+	a.SetDefault(&c.DString, "defString")
+	a.SetDefault(&c.DBool, true)
+	a.SetDefault(&c.DInt, 42)
+}
+
+type Get struct{}
+type GetArgs struct{}
+type GetState struct {
+	Config string `pulumi:"config"`
+}
+
+func (*Get) Create(ctx p.Context, name string, input GetArgs, preview bool) (string, GetState, error) {
+	config := infer.GetConfig[Config](ctx)
+	bytes, err := json.Marshal(&config)
+	return name, GetState{Config: string(bytes)}, err
+}
```

Comment:
I don't follow why we need a copy of this here? shouldn't it just be internal to config.Provider?

Reply:
Removed.